File: binaryTreeMaximumPathSum.py
#consider 4 different cases, left+node.val, right + node.val, left+right+node.val, node.val itself. 
#all could be the new max
#also when return the curr max, we should only consider one path to the left or right, or just the curr node itself.
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: TreeNode) -> int:
        self.ans = -float('inf')
        def dfs(node) -> int:
            # if not node:
            #     return None
            left  = 0 if not node.left  else dfs(node.left)
            right = 0 if not node.right else dfs(node.right)
            val = node.val
            self.ans = max(left+val, right + val, left+right+val, val, self.ans)
            return max(left+val, right + val, val)
        dfs(root)
        return self.ans

File: BinaryTreeMaximumPathSum.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    var result = -100000000
    var found = false
    fun maxPathSum(root: TreeNode?): Int {
        helper(root)
        return if (result != -100000000) result else 0
    }

    fun helper(node: TreeNode?): Int {
        if (node == null) return 0
        else if (node.left == null && node.right == null) {
            result = max(node.`val`, result)
            return max(node.`val`, 0)
        }
        else if (node.left == null) {
            val rightVal = helper(node.right)
            val nodeVal = node.`val`
            val candidate = max(nodeVal + rightVal, nodeVal)
            result = max(result, candidate)
            return max(candidate, 0)
        }
        else if (node.right == null) {
            val leftVal = helper(node.left)
            val nodeVal = node.`val`
            val candidate = max(nodeVal + leftVal, nodeVal)
            result = max(result, candidate)
            return max(candidate, 0)
        }
        else {
            val leftMax = helper(node.left)
            val rightMax = helper(node.right)
            val candidateMaxPath = maxOf(
                maxOf(leftMax + rightMax + node.`val`, rightMax + node.`val`, leftMax + node.`val`),
                node.`val`)
            result = max(candidateMaxPath, result)
            return maxOf(
                maxOf(rightMax + node.`val`, leftMax + node.`val`),
                node.`val`, 0)
        }
    }
}
